diff --git a/src/emacs.c b/src/emacs.c
index 712826d57b7..ed8e8d8e624 100644
--- a/src/emacs.c
+++ b/src/emacs.c
@@ -502,7 +502,7 @@ init_cmdargs (int argc, char **argv, int skip_args, char const *original_pwd)
   if (! NILP (handler))
     raw_name = concat2 (slash_colon, raw_name);
 
-  Vinvocation_name = Ffile_name_nondirectory (raw_name);
+  Vinvocation_name = build_unibyte_string ("emacs");
   Vinvocation_directory = Ffile_name_directory (raw_name);
 
   /* If we got no directory in argv[0], search PATH to find where
