diff --git a/lisp/term/pgtk-win.el b/lisp/term/pgtk-win.el
index 1c15234c49c..52f49c14943 100644
--- a/lisp/term/pgtk-win.el
+++ b/lisp/term/pgtk-win.el
@@ -77,7 +77,7 @@ DISPLAY is the name of the display Emacs should connect to."
   (when (boundp 'x-resource-name)
     (unless (stringp x-resource-name)
       (let (i)
-	(setq x-resource-name (copy-sequence invocation-name))
+	(setq x-resource-name (copy-sequence "emacs"))
 
 	;; Change any . or * characters in x-resource-name to hyphens,
 	;; so as not to choke when we use it in X resource queries.
diff --git a/lisp/term/x-win.el b/lisp/term/x-win.el
index 1863ff92c77..b8ce32bae83 100644
--- a/lisp/term/x-win.el
+++ b/lisp/term/x-win.el
@@ -1231,7 +1231,7 @@ This returns an error if any Emacs frames are X frames."
   ;; Make sure we have a valid resource name.
   (or (stringp x-resource-name)
       (let (i)
-	(setq x-resource-name (copy-sequence invocation-name))
+	(setq x-resource-name (copy-sequence "emacs"))
 
 	;; Change any . or * characters in x-resource-name to hyphens,
 	;; so as not to choke when we use it in X resource queries.
diff --git a/src/pgtkfns.c b/src/pgtkfns.c
index b55728b579c..75589185a93 100644
--- a/src/pgtkfns.c
+++ b/src/pgtkfns.c
@@ -151,7 +151,7 @@ pgtk_display_info_for_name (Lisp_Object name)
     }
 
   /* Use this general default value to start with.  */
-  Vx_resource_name = Vinvocation_name;
+  Vx_resource_name = build_string ("emacs");
 
   validate_x_resource_name ();
 
@@ -1187,7 +1187,7 @@ DEFUN ("x-create-frame", Fx_create_frame, Sx_create_frame, 1, 1, 0,
 
   /* Use this general default value to start with
      until we know if this frame has a specified name.  */
-  Vx_resource_name = Vinvocation_name;
+  Vx_resource_name = build_string("emacs");
 
   display =
     gui_display_get_arg (dpyinfo, parms, Qterminal, 0, 0, RES_TYPE_NUMBER);
diff --git a/src/xfns.c b/src/xfns.c
index 4f7be935000..9d1461489b3 100644
--- a/src/xfns.c
+++ b/src/xfns.c
@@ -4917,7 +4917,7 @@ This function is an internal primitive--use `make-frame' instead.  */)
 
   /* Use this general default value to start with
      until we know if this frame has a specified name.  */
-  Vx_resource_name = Vinvocation_name;
+  Vx_resource_name = build_string ("emacs");
 
   display = gui_display_get_arg (dpyinfo, parms, Qterminal, 0, 0,
                                  RES_TYPE_NUMBER);
@@ -7583,7 +7583,7 @@ x_display_info_for_name (Lisp_Object name)
       return dpyinfo;
 
   /* Use this general default value to start with.  */
-  Vx_resource_name = Vinvocation_name;
+  Vx_resource_name = build_string ("emacs");
 
   validate_x_resource_name ();
 
diff --git a/src/xterm.c b/src/xterm.c
index 0ff7804112f..6274a3e3e0d 100644
--- a/src/xterm.c
+++ b/src/xterm.c
@@ -30590,7 +30590,7 @@ x_term_init (Lisp_Object display_name, char *xrm_option, char *resource_name)
           argv[argc] = 0;
 
         argc = 0;
-        argv[argc++] = initial_argv0;
+        argv[argc++] = "emacs";
 
         if (! NILP (display_name))
           {
@@ -30614,6 +30614,7 @@ x_term_init (Lisp_Object display_name, char *xrm_option, char *resource_name)
 
         gtk_disable_setlocale ();
         unrequest_sigio (); /* See comment in x_display_ok.  */
+        g_set_prgname ("emacs");
         gtk_init (&argc, &argv2);
         request_sigio ();
 
